aleph-client/src/contract/event.rs

```diff
@@ -0,0 +1,130 @@
+//!
+//! let conn_copy = conn.clone();
+//! let contract1_copy = contract1.clone();
+//! let contract2_copy = contract2.clone();
+//!
+//! let (tx, mut rx) = unbounded();
+//! let listen = || async move {
+//!     listen_contract_events(&conn, &[contract1_copy.as_ref(), contract2_copy.as_ref()], tx).await?;
+//!     <Result<(), anyhow::Error>>::Ok(())
+//! };
+//! let join = tokio::spawn(listen());
+//!
+//! contract1.contract_exec0(&signed_conn, "some_method").await?;
+//! contract2.contract_exec0(&signed_conn, "some_other_method").await?;
+//!
+//! println!("Received event {:?}", rx.next().await);
+//!
+//! rx.close();
+//! join.await??;
+//!
+//! #   Ok(())
+//! # }
+//! ```
+
+use std::collections::HashMap;
+
+use anyhow::{bail, Result};
+use contract_transcode::Value;
+use futures::{channel::mpsc::UnboundedSender, StreamExt};
+
+use crate::{contract::ContractInstance, AccountId, Connection};
+
+/// Represents a single event emitted by a contract.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ContractEvent {
+    /// The address of the contract that emitted the event.
+    pub contract: AccountId,
+    /// The name of the event.
+    pub ident: Option<String>,
+    /// Data contained in the event.
+    pub data: HashMap<String, Value>,
+}
+
+/// Starts an event listening loop.
+///
+/// Will send contract event and every error encountered while fetching through the provided [UnboundedSender].
+/// Only events coming from the address of one of the `contracts` will be decoded.
+///
+/// The loop will terminate once `sender` is closed.
+pub async fn listen_contract_events(
+    conn: &Connection,
+    contracts: &[&ContractInstance],
+    sender: UnboundedSender<Result<ContractEvent>>,
+) -> Result<()> {
+    let mut block_subscription = conn.client.blocks().subscribe_finalized().await?;
+
+    while let Some(block) = block_subscription.next().await {
+        if sender.is_closed() {
+            break;
+        }
+
+        let block = block?;
+
+        for event in block.events().await?.iter() {
+            let event = event?;
+
+            if let Some(event) =
+                event.as_event::<crate::api::contracts::events::ContractEmitted>()?
+            {
+                if let Some(contract) = contracts
+                    .iter()
+                    .find(|contract| contract.address() == &event.contract)
+                {
+                    let data = zero_prefixed(&event.data);
+                    let event = contract
+                        .transcoder
+                        .decode_contract_event(&mut data.as_slice());
+
+                    sender.unbounded_send(build_event(contract.address().clone(), event))?;
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// The contract transcoder assumes there is an extra byte (that it discards) indicating the size of the data. However,
+/// data arriving through the subscription as used in this file don't have this extra byte. This function adds it.
```

Comment:
```suggestion
/// data arriving through the subscription as used in this file doesn't have this extra byte. This function adds it.
```

Reply:
I think both are correct, data can be plural or singular. I guess `don't` is more formal than `doesn't`.

Reply:
Right, didn't know that :)